### DISCUSSION TITLE
Creating a DeviceObject

### DISCUSSION THREAD

**Discussion by nsnyder1992** [ChristianTremblay/BAC0#365] [Q&A]
I am trying to add a DeviceObject (bacpypes.object.DeviceObject) to my BAC0 app, with the localDate and localTime properties. I need these properties to recreate devices that have these attributes, to test against my bacnet driver I am writing in java.

 I tried to use the ObjectFactory like so:

`localDate = ObjectFactory(
        DeviceObject,
        56,
        "localDate",
        presentValue=now.strftime("%x"),
        description="Local Date",
    )`

But of course get a KeyError("Unknown"), because there is no presentValue property for the DeviceObject type. 

Is DeviceObject type a multistate value? If so, should I create my own version of the multistate(**kwargs) function in  BAC0.core.devices.local.models? 

It would be greatly appreciated if someone could point me in the right direction. :)

**Comment by kheldaroz402** [comment]
the DeviceObject is made up a lot of different properties
e.g.
![2022-11-30 15_04_33-REDLINE_ASHRAE_135_2020-2016 pdf - Foxit PDF Reader](https://user-images.githubusercontent.com/55415109/204705216-323903e6-ee57-4d48-a423-32a99739936a.png)

some of which are arrays etc.

the date time properties are defined as

```
12.11.23 Local_Time
The Local_Time property, of type Time, shall indicate the time of day to the best of the device's
knowledge. If the BACnet device does not have any knowledge of time or date, then the Local_Time
property shall be omitted. This property shall be present if the BACnet device is capable of tracking
date and time. If the device restarts, and there is no local time available yet, then the clock shall be
initialized to the time 00:00:00.00 and the device shall commence tracking the time.
12.11.24 Local_Date
The Local_Date property, of type Date, shall indicate the date to the best of the device's knowledge.
If the BACnet device does not have any knowledge of time or date, then the Local_Date property
shall be omitted. This property shall be present if the BACnet device is capable of tracking date and
time. If the device restarts, and there is no local date available yet, then the clock shall be initialized
to a date on or before January 1, 1990 and the device shall commence tracking the date.
12.11.25 UTC_Offset
The UTC_Offset property, of type INTEGER, shall indicate the number of minutes (-1440 to +1440)
offset between local standard time and Universal Time Coordinated. The time zones to the west of
the zero degree meridian shall be positive values, and those to the east shall be negative values. The
value of the UTC_Offset property is subtracted from the UTC received in UTCTimeSynchronization
service requests to calculate the correct local standard time. UTC_Offset shall be configurable and
accept any multiple of 15 minutes across the full range.
12.11.26 Daylight_Savings_Status
The Daylight_Savings_Status property, of type BOOLEAN, shall indicate whether daylight savings
time is in effect (TRUE) or not (FALSE) at the BACnet device's location.
```

which is the official definition.

however I don't know if BAC0 supports these as they are optional depending on the features used

**Comment by ChristianTremblay** [comment]
The async version of BAC0 now depends on bacpypes3 and the device object is way more complete. You will be able to add localTime, etc...
